src/java/org/apache/cassandra/index/sai/memory/TrieMemoryIndex.java

```diff
@@ -252,13 +252,13 @@ private void setMinMaxTerm(ByteBuffer term)
 
     private ByteComparable encode(ByteBuffer input)
     {
-        return indexContext.isLiteral() ? version -> ByteSource.append(ByteSource.of(input, version), ByteSource.TERMINATOR)
-                                        : version -> TypeUtil.asComparableBytes(input, indexContext.getValidator(), version);
+        return indexContext.isLiteral() ? version -> ByteSource.append(ByteSource.of(input, ByteComparable.Version.OSS41), ByteSource.TERMINATOR)
```

Comment:
This threw me off. That is, I got immediately that the goal was to force the old version of `ByteComparable`, but I didn't get why this did it this way, ignoring the provided `version`. And this feels immediately error prone to me, because if I'm using a `ByteComparable`, I kind of expect `asComparableBytes(version)` to give me a byte-source encoded in the version I asked it.

Looking closer, I'm understanding that it is done this way because `Trie` hard-codes the version it uses, with `BYTE_COMPARABLE_VERSION` (some method like `ClusteringComparator.clusteringFromByteComparable`, which is used by SAI, also independently hard-code that version as it happens), and so my first reaction was to suggest making the `version` be a property of a `Trie`.

But I'm realizing that the issue (as I perceive it) is more generic than that. It seems there is plenty of places where code returns a `ByteComparable`, yet it's not able to produce anything else than the underlying encoded version, That's the case of every `ByteComparable` returned by a `Trie` in particular. I'm very very worried about this making it extremely easy to create bugs moving forward.

Case in point, I think there is an unaddressed issue with SAI: the code in `RowAwarePrimaryKeyMap#supplier` clearly  expects `OSS41` version and `peekable` will be in that version in particular, but it also has:
```java
Clustering clustering = clusteringComparator.size() == 0
                        ? Clustering.EMPTY
                        : clusteringComparator.clusteringFromByteComparable(ByteBufferAccessor.instance,
                                                                            v -> ByteSourceInverse.nextComponentSource(peekable));
```
and as pointed above, `clusteringFromByteComparable` hard-codes internally its `version` to `OSS50`. So when that method, for a clustering component type, calls `subtype(cc).fromComparableBytes(accessor, orderedBytes, version)`, the `orderedBytes` will effectively be encoded with `OSS41` but `version` will be `OSS50`, so for a type like `IntegerType` where those differs, I believe this will do the wrong thing.

Re-SAI, I'll note that the code actually kind of attempt to make the `ByteComparable.Version` dependent on the index version (SAI `Version` class has a `byteComparableVersionFor` method), so hard-coding the version is probably incorrect here too, but I'll admit that the SAI is a bit messy in that regard because there is still quite a bunch of hard-coded `ByteComparable.Version.OSS41` here and there.

Anyway, to sum this up, I think there is a concrete bug here for SAI, but I'm also rather worried about the use of `ByteComparable` where the `version` is essentially ignored. But I understand the later is not necessarily new to this patch, so I'm happy to leave cleaning that up to a followup if we prefer, as long as we at least fix the former here. In fact, I don't mind if we want to move the "supports OSS50 format" part into a separate PR completely.

But fwiw, my suggestion would be to start by trying to avoid hard-coding the version as much as possible. I do feel the `version` should be a property of the `Trie` even if that doesn't quite solve the problem of the `ByteComparable` returned when querying the trie (but maybe the trie could wrap what it returns to at least validate that when a `version` is provided, then it's the same as the one of the Trie and either fail when that's not the case or, if we're really motivated, re-interpret into the requested version). And SAI should be cleaned up to always use the version that is attached to it's "index version" (which would allows to also move to the OSS50 version with the next index version if we want).

Reply:
> to supply each returned value with its version and throw an error on mismatch
> Or maybe we should bite the bullet and make another type of byte container to return?

Yeah, that's the 2 options that crossed my mind. I "think" that adding a new type may be the cleaner option long term and I would have a small preference for that in theory, but I'm also not sure how much work that is and "checking if the versions match" would avoid problems in practice, so I'm ok with that if that's a lot simpler/you prefer it.